File: MyExamples/TestServiceApp/app/src/main/java/examples/my/android/testserviceapp/MainActivity.java
package examples.my.android.testserviceapp;

import android.content.Intent;
import android.content.IntentFilter;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;

public class MainActivity extends AppCompatActivity {

    private Button mStartServiceBtn;
    private Button mStopServiceBtn;
    private Button mSendBroadcastBtn;
    private SimpleReceiver mSimpleReceiver;
    private IntentFilter mIntentFilter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mStartServiceBtn = (Button) findViewById(R.id.btn_start_service);
        mStopServiceBtn = (Button) findViewById(R.id.btn_stop_service);
        mSendBroadcastBtn = (Button) findViewById(R.id.btn_send_broadcast);

        mStartServiceBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(MainActivity.this, TestService.class);
                startService(intent);
            }
        });

        mStopServiceBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(MainActivity.this, TestService.class);
                stopService(intent);
            }
        });

        mSendBroadcastBtn.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View v) {
                sendBroadcast(new Intent(SimpleReceiver.SIMPLE_ACTION));
            }
        });

        mSimpleReceiver = new SimpleReceiver();
        mIntentFilter = new IntentFilter(SimpleReceiver.SIMPLE_ACTION);
    }

    @Override
    protected void onResume() {
        super.onResume();
        registerReceiver(mSimpleReceiver, mIntentFilter);
    }

    @Override
    protected void onPause() {
        super.onPause();
        unregisterReceiver(mSimpleReceiver);
    }
}

File: MyExamples/PermissionsExample/app/src/main/java/my/example/permissionsexample/MainActivity.kt
package my.example.permissionsexample

import android.Manifest
import android.content.pm.PackageManager
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v7.app.AlertDialog
import android.text.TextUtils
import android.widget.Button
import android.widget.EditText

class MainActivity : AppCompatActivity() {

    companion object {
        const val WRITE_PERMISSION_RC = 1
    }

    private lateinit var mEditText: EditText
    private lateinit var mSaveButton: Button

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        mEditText = findViewById<EditText>(R.id.et_write)
        mSaveButton = findViewById<Button>(R.id.btn_save)

        mSaveButton.setOnClickListener {
            val text = mEditText.text.toString()
            saveTextToFile(text)

        }
    }

    private fun saveTextToFile(text: String?) {
        if (TextUtils.isEmpty(text)) {
            toast("EditText пустой!")
        } else {
            requestPermissionAndSave(text)
        }
    }

    private fun requestPermissionAndSave(text: String?) {
        if (isWritePermissionGranted()) {
            writeToFile(text)
        } else {
            requestWritePermission()
        }
    }

    private fun requestWritePermission() {
        if (ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)) {

            //show
            AlertDialog.Builder(this)
                .setMessage("Без этого разрешения работать не будет!")
                .setPositiveButton("Понятно") { dialog, which ->

                    ActivityCompat.requestPermissions(
                        this@MainActivity,
                        arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
                        WRITE_PERMISSION_RC
                    )

                }.show()

        } else {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
                WRITE_PERMISSION_RC
            )
        }
    }

    private fun writeToFile(text: String?) {
        toast("$text записано в файл!")
    }

    private fun isWritePermissionGranted(): Boolean {
        return ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) ==
                PackageManager.PERMISSION_GRANTED
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        if (requestCode != WRITE_PERMISSION_RC) return
        if (grantResults.size != 1) return

        if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            val text = mEditText.text.toString()
            saveTextToFile(text)
        } else {
            AlertDialog.Builder(this)
                .setMessage("Вы можете предоставить разрешение в настройках устройства.")
                .setPositiveButton("Понятно", null)
                .show()
        }
    }
}
